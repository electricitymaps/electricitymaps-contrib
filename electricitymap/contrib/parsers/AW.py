#!/usr/bin/env python3

from datetime import datetime
from logging import Logger, getLogger
from typing import Any
from zoneinfo import ZoneInfo

from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey

from .lib.exceptions import ParserException

FUEL_MAPPING = {
    "Fossil": "oil",
    "Wind": "wind",
    "TotalSolar": "solar",
    "total_bio_gas": "biomass",
}
TIMEZONE = ZoneInfo("America/Aruba")
PRODUCTION_URL = (
    "https://www.webaruba.com/renewable-energy-dashboard/app/rest/results.json"
)
SOURCE = "webaruba.com"


def fetch_production(
    zone_key: ZoneKey = ZoneKey("AW"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    r = session or Session()
    # User agent is mandatory or services answers 404
    headers = {
        "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko)"
    }
    response = r.get(PRODUCTION_URL, headers=headers)
    aruba_json = response.json()
    top_data = aruba_json["dashboard_top_data"]
    mix = ProductionMix()
    sources_total = 0
    for fuel, mode in FUEL_MAPPING.items():
        value = float(top_data[fuel]["value"])
        mix.add_value(mode, value)
        sources_total += value
    # "unknown" is when data reported in the categories above is less than total reported.
    # If categories sum up to more than total, accept the datapoint, but only if it's less than 10% of total.
    # This helps avoid missing data when it's a little bit off, due to rounding or reporting
    reported_total = float(top_data["TotalPower"]["value"])

    if (sources_total / reported_total) > 1.1:
        raise ParserException(
            "AW.py",
            f"AW parser reports fuel sources add up to {sources_total} but total generation {reported_total} is lower",
            zone_key,
        )

    mix.add_value(
        "unknown", reported_total - sources_total, correct_negative_with_zero=True
    )
    # We're using Fossil data to get timestamp in correct time zone
    local_date_time = datetime.strptime(
        top_data["total_bio_gas"]["timestamp"], "%Y-%m-%d %H:%M:%S.%f"
    ).replace(tzinfo=TIMEZONE)

    production_list = ProductionBreakdownList(logger)
    production_list.append(
        zone_key,
        local_date_time,
        SOURCE,
        production=mix,
    )

    return production_list.to_list()


if __name__ == "__main__":
    print(fetch_production())
