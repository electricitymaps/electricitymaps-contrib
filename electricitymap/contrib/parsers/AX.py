from datetime import datetime, timedelta
from logging import Logger, getLogger
from re import findall
from zoneinfo import ZoneInfo

from bs4 import BeautifulSoup
from requests import Response, Session

from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    ProductionBreakdownList,
    TotalConsumptionList,
)
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey

from .lib.exceptions import ParserException

IFRAME_URL = "https://grafik.kraftnat.ax/grafer/tot_inm_24h_15.php"
TIME_ZONE = ZoneInfo("Europe/Mariehamn")
SOURCE = "kraftnat.ax"


def fetch_data(session: Session, logger: Logger):
    """Fetch data from the iFrame."""

    response: Response = session.get(IFRAME_URL)
    response.raise_for_status()

    soup = BeautifulSoup(response.text, "html.parser").find_all("script")
    result_time_series = findall(r"data: \[(.*?)\]", str(soup))
    if len(result_time_series) != 3:
        raise ParserException(
            "AX.py",
            "Did not find the expected amount of results. Check if the website has changed.",
        )
    time_series: list = result_time_series[0].split(",")
    raw_data: list[str] = findall(r"data:\[(.*?)\]", str(soup))
    if len(raw_data) != 6:
        raise ParserException(
            "AX.py",
            "The raw data did not match the expected format. Check if the website has changed.",
        )
    for raw in raw_data:
        if len(raw.split(",")) != len(time_series):
            raise ParserException(
                "AX.py",
                "The raw data did not match the length of the the time series. Check if the website has changed.",
            )
    data_list = []
    for time, sweden, alink, fossil, gustavs, wind, consumption in zip(
        time_series,
        raw_data[0].split(","),
        raw_data[1].split(","),
        raw_data[2].split(","),
        raw_data[3].split(","),
        raw_data[4].split(","),
        raw_data[5].split(","),
        strict=True,
    ):
        data_list.append(
            {
                "time": str(time.replace('"', "")),
                "sweden": float(sweden),
                "alink": float(alink),
                "fossil": float(fossil),
                "gustavs": float(gustavs),
                "wind": float(wind),
                "consumption": float(consumption),
            }
        )
    return data_list


def formatted_data(
    zone_key: ZoneKey | None,
    zone_key1: ZoneKey | None,
    zone_key2: ZoneKey | None,
    session: Session,
    logger: Logger,
    data_type: str,
):
    """Format data to Electricity Map standards."""
    data_list = fetch_data(session, logger)
    data_list.reverse()
    date_time = datetime.now(TIME_ZONE)
    date = date_time.replace(
        hour=int(data_list[0]["time"].split(":")[0]),
        minute=int(data_list[0]["time"].split(":")[1]),
        second=0,
        microsecond=0,
    )
    if date > date_time:
        date = date - timedelta(days=1)
    exchanges = ExchangeList(logger)
    production_breakdowns = ProductionBreakdownList(logger)
    consumption = TotalConsumptionList(logger)
    for data in data_list:
        corrected_date = date - timedelta(minutes=15 * data_list.index(data))
        if data_type == "production" and zone_key is not None:
            production_mix = ProductionMix(wind=data["wind"], oil=data["fossil"])
            production_breakdowns.append(
                datetime=corrected_date,
                production=production_mix,
                source=SOURCE,
                zoneKey=zone_key,
            )
        elif data_type == "consumption" and zone_key is not None:
            consumption.append(
                datetime=corrected_date,
                consumption=data["consumption"],
                source=SOURCE,
                zoneKey=zone_key,
            )
        elif data_type == "exchange":
            if zone_key1 == ZoneKey("AX") and zone_key2 == ZoneKey("SE-SE3"):
                exchanges.append(
                    datetime=corrected_date,
                    netFlow=data["sweden"] * -1,
                    source=SOURCE,
                    zoneKey=ZoneKey("AX->SE-SE3"),
                )
            elif zone_key1 == ZoneKey("AX") and zone_key2 == ZoneKey("FI"):
                exchanges.append(
                    datetime=corrected_date,
                    netFlow=round(data["alink"] + data["gustavs"], 2) * -1,
                    source=SOURCE,
                    zoneKey=ZoneKey("AX->FI"),
                )
            else:
                raise ParserException(
                    "AX.py",
                    "This parser can only fetch data between Åland <-> Sweden and Åland <-> Finland",
                )
        else:
            raise ParserException(
                "AX.py",
                "The datasource currently implemented is only for production, consumption and exchange data",
                zone_key,
            )

    if data_type == "production":
        return production_breakdowns.to_list()
    elif data_type == "consumption":
        return consumption.to_list()
    elif data_type == "exchange":
        return exchanges.to_list()
    else:
        raise ParserException(
            "AX.py",
            "The datasource currently implemented is only for production, consumption and exchange data",
            zone_key,
        )


def fetch_production(
    zone_key: ZoneKey = ZoneKey("AX"),
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Fetch production data."""

    if target_datetime is not None:
        raise ParserException(
            "AX.py", "The datasource currently implemented is only for real time data"
        )

    return formatted_data(
        zone_key=zone_key,
        zone_key1=None,
        zone_key2=None,
        session=session,
        logger=logger,
        data_type="production",
    )


def fetch_consumption(
    zone_key: ZoneKey = ZoneKey("AX"),
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Fetch consumption data."""

    if target_datetime is not None:
        raise ParserException(
            "AX.py", "The datasource currently implemented is only for real time data"
        )

    return formatted_data(
        zone_key=zone_key,
        zone_key1=None,
        zone_key2=None,
        session=session,
        logger=logger,
        data_type="consumption",
    )


def fetch_exchange(
    zone_key1: ZoneKey,
    zone_key2: ZoneKey,
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Fetch exchange data."""

    if target_datetime is not None:
        raise ParserException(
            "AX.py", "The datasource currently implemented is only for real time data"
        )

    return formatted_data(
        zone_key=None,
        zone_key1=zone_key1,
        zone_key2=zone_key2,
        session=session,
        logger=logger,
        data_type="exchange",
    )
