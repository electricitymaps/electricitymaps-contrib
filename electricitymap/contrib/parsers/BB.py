from datetime import datetime, timedelta
from logging import Logger, getLogger
from typing import Any
from zoneinfo import ZoneInfo

import pandas as pd

# The request library is used to fetch content through HTTP
import requests
from requests import Session
from requests.auth import AuthBase

from electricitymap.contrib.lib.models.event_lists import (
    ProductionBreakdownList,
    TotalConsumptionList,
)
from electricitymap.contrib.lib.models.events import ProductionMix, StorageMix
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.config import refetch_frequency
from electricitymap.contrib.parsers.lib.exceptions import ParserException

SOURCE = "energy.gov.bb"
base_url = "https://dataservices.acelerex.com/api"


class JwtPublicAuth(AuthBase):
    def __init__(self):
        self.token = None

    def __call__(self, r):
        if self.token is None:
            self.token = self.get_token()
        r.headers["Authorization"] = f"Bearer {self.token}"
        return r

    def get_token(self) -> str:
        url = base_url + "/authentication"
        response = requests.post(
            url, json={"dashboardId": "barbados", "strategy": "jwt-public"}
        )
        return response.json()["accessToken"]


def session_with_auth(session: Session) -> Session:
    if session.auth is None:
        session.auth = JwtPublicAuth()
    return session


def parse_operational_data(response_json) -> tuple[list[str], list[list[str]]]:
    header, *table = (x for x in response_json["data"][0]["data"]["data"] if len(x) > 1)
    return header, table


def fetch_operational_data(
    zone_key,
    session: Session,
    target_datetime: datetime | None,
    logger: Logger,
) -> pd.DataFrame:
    session = session_with_auth(session)

    params = {"name": "operational-data", "$sort[created_at]": -1}

    header = None
    table = None

    if target_datetime is None:
        data = session.get(base_url + "/barbados/uploads", params=params).json()
        header, table = parse_operational_data(data)
    elif target_datetime > datetime(year=2023, month=2, day=27):
        expected_date = target_datetime.astimezone(
            ZoneInfo("America/Barbados")
        ).strftime("%Y-%m-%d")

        for i in range(10):
            url_date = (
                (target_datetime + timedelta(days=i))
                .astimezone(ZoneInfo("America/Barbados"))
                .strftime("%Y-%m-%d")
            )
            params["created_at[$lt]"] = url_date
            data = session.get(base_url + "/barbados/uploads", params=params).json()
            header, table = parse_operational_data(data)
            if table[0][0][:10] == expected_date:
                break
            if table[0][0][:10] > expected_date:
                raise ParserException("BB.py", "Date not found", zone_key)
    else:
        raise ParserException(
            "BB.py",
            "This parser is not yet able to parse dates before 2023-02-27",
            zone_key,
        )

    if table is None or header is None:
        raise ParserException("BB.py", "No data found", zone_key)

    operational_data = pd.DataFrame(
        table,
        columns=header,
    )
    columns = [
        "24 hr monitored solar production",
        "24 hr monitored wind production",
        "24 hr storage charging",
        "24 hr storage discharging",
        "24 hr fossil production",
        "24 hr Net demand",
        "24 hr Curtailment",
    ]
    for columm in columns:
        operational_data[columm] = pd.to_numeric(operational_data[columm])
    return operational_data


@refetch_frequency(timedelta(days=1))
def fetch_production(
    zone_key: ZoneKey = ZoneKey("BB"),
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict] | dict:
    operational_data = fetch_operational_data(
        zone_key, session, target_datetime, logger
    )

    production_list = ProductionBreakdownList(logger=logger)

    for _index, row in operational_data.iterrows():
        production_list.append(
            zoneKey=zone_key,
            datetime=datetime.fromisoformat(row["datetime"]),
            production=ProductionMix(
                solar=row["24 hr monitored solar production"],
                wind=row["24 hr monitored wind production"],
                oil=row["24 hr fossil production"],
            ),
            storage=StorageMix(
                battery=(
                    row["24 hr storage discharging"] + row["24 hr storage charging"]
                )
                * -1
            ),
            source=SOURCE,
        )
    return production_list.to_list()


@refetch_frequency(timedelta(days=1))
def fetch_consumption(
    zone_key: ZoneKey,
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> dict[str, Any] | list[dict[str, Any]]:
    operational_data = fetch_operational_data(
        zone_key, session, target_datetime, logger
    )

    consumption_list = TotalConsumptionList(logger=logger)

    for _index, row in operational_data.iterrows():
        consumption_list.append(
            zoneKey=zone_key,
            datetime=datetime.fromisoformat(row["datetime"]),
            consumption=row["24 hr Net demand"],
            source=SOURCE,
        )
    return consumption_list.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Maps backend, but handy for testing."""

    print("fetch_production(XX) ->")
    print(fetch_production(ZoneKey("BB")))
    fetch_production(ZoneKey("BB"), target_datetime=datetime(2024, 2, 2))
    print("fetch_consumption(XX) ->")
    print(fetch_consumption(ZoneKey("BB")))
