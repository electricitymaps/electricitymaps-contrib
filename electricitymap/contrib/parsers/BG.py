from datetime import datetime, timezone
from logging import Logger, getLogger

from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.exceptions import ParserException

PARSER = "BG.py"

SOURCE = "eso.bg"
SOURCE_API_URL = "http://www.eso.bg/api/rabota_na_EEC_json.php?en"

PRODUCTION_TYPE_TO_PRODUCTION_MODE = {
    "NPP": "nuclear",
    "CHP": "coal",
    "Heating TPPs": "gas",
    "Factory TPPs": "gas",
    "HPP": "hydro",
    "Small HPPs": "hydro",
    "Wind": "wind",
    "PV": "solar",
    "Bio": "biomass",
}


def fetch_production(
    zone_key: ZoneKey = ZoneKey("BG"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    session = session or Session()

    if target_datetime is not None:
        raise ParserException(
            PARSER, "This parser is not yet able to parse historical data", zone_key
        )

    time = datetime.now(timezone.utc)
    response = session.get(SOURCE_API_URL)
    if not response.ok:
        raise ParserException(
            PARSER,
            f"Exception when fetching production error code: {response.status_code}: {response.text}",
            zone_key,
        )

    response_payload = response.json()
    logger.debug(f"Raw generation breakdown: {response_payload}")

    production_mix = ProductionMix()
    for header, value in response_payload:
        production_type, _delimiter, _percentage = header.rpartition(" ")
        production_mode = PRODUCTION_TYPE_TO_PRODUCTION_MODE[production_type]
        production_mix.add_value(production_mode, value)

    production_breakdown_list = ProductionBreakdownList(logger)
    production_breakdown_list.append(
        zoneKey=zone_key,
        datetime=time,
        source=SOURCE,
        production=production_mix,
    )
    return production_breakdown_list.to_list()
