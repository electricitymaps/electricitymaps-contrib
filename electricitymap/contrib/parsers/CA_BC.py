#!/usr/bin/env python3

from datetime import datetime, timedelta
from logging import Logger, getLogger
from typing import Any
from zoneinfo import ZoneInfo

import pandas as pd
from requests import Response, Session

from electricitymap.contrib.config import ZONES_CONFIG
from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    TotalConsumptionList,
)
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.config import refetch_frequency
from electricitymap.contrib.parsers.lib.exceptions import ParserException

# More info:
# https://www.bchydro.com/energy-in-bc/our_system/transmission/transmission-system/actual-flow-data.html

HISTORICAL_LOAD_REPORTS = "https://www.bchydro.com/content/dam/BCHydro/customer-portal/documents/corporate/suppliers/transmission-system/balancing_authority_load_data/Historical%20Transmission%20Data/BalancingAuthorityLoad{0}.xls"
EXCHANGES_URL = "https://www.bchydro.com/bctc/system_cms/actual_flow/latest_values.txt"
TIMEZONE = ZoneInfo(ZONES_CONFIG.get(ZoneKey("CA-BC"), {}).get("timezone"))
SOURCE = "bchydro.com"
PUBLICATION_DELAY = timedelta(days=31)

EXCHANGE_POSITION_MULTIPLIER = {
    ZoneKey("CA-BC->US-NW-BPAT"): (1, 1),
    ZoneKey("CA-AB->CA-BC"): (2, -1),
}


@refetch_frequency(timedelta(days=1))
def fetch_consumption(
    zone_key: ZoneKey,
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
):
    r = session or Session()
    consumption_list = TotalConsumptionList(logger)
    if target_datetime is None:
        start_date = datetime.now(tz=TIMEZONE) - PUBLICATION_DELAY - timedelta(days=1)
        end_date = datetime.now(tz=TIMEZONE)
    else:
        start_date = target_datetime - timedelta(days=1)
        end_date = target_datetime
    # If the range spans multiple years, we need to fetch each year separately.
    for year in range(start_date.year, end_date.year + 1):
        url = HISTORICAL_LOAD_REPORTS.format(year)
        response: Response = r.get(url)
        if not response.ok:
            raise ParserException(
                "CA_BC.py",
                f"Could not fetch load report for year {year}",
                zone_key,
            )
        df = pd.read_excel(response.content, skiprows=3)
        df = df.rename(columns={"Date ?": "Date"})
        # The first hour of each day is set at 1 and the last at 24 so we need to subtract 1
        # to correctly align dates and hours.
        # Due to DST there might be some missing hours, we use shift_backward to fill them with the previous hour.
        df["datetime"] = pd.to_datetime(
            df["Date"] + " " + (df["HE"] - 1).astype("str"), format="%m/%d/%Y %H"
        ).dt.tz_localize(TIMEZONE, nonexistent="shift_backward")
        df = df.set_index("datetime")
        selected_times = df[start_date:end_date]
        for row in selected_times.iterrows():
            consumption_list.append(
                zoneKey=zone_key,
                datetime=row[0].to_pydatetime(),
                source=SOURCE,
                consumption=row[1]["Control Area Load"],
            )
    return consumption_list.to_list()


def fetch_exchange(
    zone_key1: str,
    zone_key2: str,
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Requests the last known power exchange (in MW) between two countries."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    r = session or Session()
    response = r.get(EXCHANGES_URL)
    obj = response.text.split("\r\n")[1].replace("\r", "").split(",")

    parsed_datetime = datetime.strptime(obj[0], "%d-%b-%y %H:%M:%S").replace(
        tzinfo=TIMEZONE
    )

    sortedZoneKeys = ZoneKey("->".join(sorted([zone_key1, zone_key2])))
    if sortedZoneKeys not in EXCHANGE_POSITION_MULTIPLIER:
        raise NotImplementedError("This exchange pair is not implemented")
    position, multiplier = EXCHANGE_POSITION_MULTIPLIER[sortedZoneKeys]

    exchanges = ExchangeList(logger)
    exchanges.append(
        sortedZoneKeys,
        parsed_datetime,
        SOURCE,
        float(obj[position]) * multiplier,
    )
    return exchanges.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""
    print("fetch_exchange(CA-AB, CA-BC) ->")
    print(fetch_exchange("CA-AB", "CA-BC"))
    print("fetch_consumption(CA-BC) ->")
    print(
        fetch_consumption(
            ZoneKey("CA-BC"), target_datetime=datetime(2023, 3, 1, 0, 0, 0, 0, TIMEZONE)
        )
    )
