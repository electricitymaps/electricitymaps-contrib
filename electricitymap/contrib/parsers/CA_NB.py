#!/usr/bin/env python3

from datetime import datetime, timezone
from logging import Logger, getLogger
from typing import Any

# BeautifulSoup is used to parse HTML to get information
from bs4 import BeautifulSoup
from requests import Session

from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    ProductionBreakdownList,
)
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey

SOURCE = "tso.nbpower.com"
EXCHANGE_TO_FLOWS = {
    ZoneKey("CA-NB->CA-QC"): ["QUEBEC"],
    # all of these exports are to Maine
    # (see https://www.nbpower.com/en/about-us/our-energy/system-map/),
    # currently this is mapped to ISO-NE
    ZoneKey("CA-NB->US-NE-ISNE"): ["EMEC", "ISO-NE", "MPS"],
    ZoneKey("CA-NB->CA-NS"): ["NOVA SCOTIA"],
    ZoneKey("CA-NB->CA-PE"): ["PEI"],
}


def _get_new_brunswick_flows(requests_obj):
    """
    Gets current electricity flows in and out of New Brunswick.

    There is no reported data timestamp in the page.
    The page returns current time and says "Times at which values are sampled may vary by as much as 5 minutes."
    """

    url = "https://tso.nbpower.com/Public/en/SystemInformation_realtime.asp"
    response = requests_obj.get(url)

    soup = BeautifulSoup(response.text, "html.parser")

    table = soup.find("table", attrs={"bordercolor": "#191970"})

    rows = table.find_all("tr")

    headers = rows[1].find_all("td")
    values = rows[2].find_all("td")

    flows = {
        headers[i].text.strip(): float(row.text.strip()) for i, row in enumerate(values)
    }

    return flows


def fetch_production(
    zone_key: ZoneKey = ZoneKey("CA-NB"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Requests the last known production mix (in MW) of a given country."""

    """
    In this case, we are calculating the amount of electricity generated
    in New Brunswick, versus imported and exported elsewhere.
    """

    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    requests_obj = session or Session()
    flows = _get_new_brunswick_flows(requests_obj)

    # nb_flows['NB Demand'] is the use of electricity in NB
    # 'EMEC', 'ISO-NE', 'MPS', 'NOVA SCOTIA', 'PEI', and 'QUEBEC'
    # are exchanges - positive for exports, negative for imports
    # Electricity generated in NB is then 'NB Demand' plus all the others

    generated = (
        flows["NB Demand"]
        + flows["EMEC"]
        + flows["ISO-NE"]
        + flows["MPS"]
        + flows["NOVA SCOTIA"]
        + flows["PEI"]
        + flows["QUEBEC"]
    )
    production = ProductionBreakdownList(logger)
    production.append(
        zoneKey=zone_key,
        # Using the current utc time because the page returns the current time.
        datetime=datetime.now(tz=timezone.utc).replace(
            minute=0, second=0, microsecond=0
        ),
        source=SOURCE,
        production=ProductionMix(
            unknown=generated,
        ),
    )

    return production.to_list()


def fetch_exchange(
    zone_key1: ZoneKey,
    zone_key2: ZoneKey,
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Requests the last known power exchange (in MW) between two regions."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    sorted_zone_keys = "->".join(sorted([zone_key1, zone_key2]))

    requests_obj = session or Session()
    flows = _get_new_brunswick_flows(requests_obj)

    # In this source, positive values are exports and negative are imports.
    # In expected result, "net" represents an export.
    # So these can be used directly.

    if sorted_zone_keys not in EXCHANGE_TO_FLOWS:
        raise NotImplementedError("This exchange pair is not implemented")
    exchanges = ExchangeList(logger)
    exchanges.append(
        zoneKey=sorted_zone_keys,
        datetime=datetime.now(tz=timezone.utc).replace(
            minute=0, second=0, microsecond=0
        ),
        source=SOURCE,
        netFlow=sum([flows[flow] for flow in EXCHANGE_TO_FLOWS[sorted_zone_keys]]),
    )

    return exchanges.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())

    print('fetch_exchange("CA-NB", "CA-PE") ->')
    print(fetch_exchange(ZoneKey("CA-NB"), ZoneKey("CA-PE")))
