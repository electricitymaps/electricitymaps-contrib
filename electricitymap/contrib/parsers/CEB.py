#!/usr/bin/env python3

"""Quarter-hourly data parser for Sri Lanka
Fetches data for the previous day in 15-minute increments
Data is from the backend for the load curve graph on https://cebcare.ceb.lk/gensum/details
"""

import json
from datetime import datetime, timedelta
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

# The request library is used to fetch content through HTTP
from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.config import refetch_frequency
from electricitymap.contrib.parsers.lib.exceptions import ParserException

TIMEZONE = ZoneInfo("Asia/Colombo")
GENERATION_BREAKDOWN_URL = "https://cebcare.ceb.lk/GenSum/GetLoadCurveData"
SOURCE_NAME = "ceb.lk"

PRODUCTION_MAPPING = {
    "SPP Biomass": "biomass",
    "Coal": "coal",
    "Thermal-Oil": "oil",
    "SPP Minihydro": "hydro",
    "Major Hydro": "hydro",
    "Solar": "solar",
    "Wind": "wind",
}


@refetch_frequency(timedelta(hours=24))
def fetch_production(
    zone_key: ZoneKey = ZoneKey("LK"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
):
    """Requests the previous day's production mix (in MW) for Sri Lanka, per quarter-hour"""
    if target_datetime is None:
        target_datetime = datetime.now(tz=TIMEZONE)

    if session is None:
        session = Session()

    params = {
        "date": target_datetime.strftime("%Y-%m-%d"),
    }

    response = session.get(GENERATION_BREAKDOWN_URL, params=params)

    if not response.ok:
        raise ParserException(
            "CEB.py",
            f"Failed to fetch production data. Response Code: {response.status_code}\nError:\n{response.text}",
            zone_key,
        )

    source_data = json.loads(
        response.json()
    )  # Response is double encoded; a JSON array encoded as a JSON string

    production_data = ProductionBreakdownList(logger)

    for quarter_hourly_source_data in source_data:
        production = ProductionMix()

        for generation_type, outputInMW in quarter_hourly_source_data.items():
            if generation_type == "DateTime":
                continue
            elif generation_type in PRODUCTION_MAPPING:
                production.add_value(PRODUCTION_MAPPING[generation_type], outputInMW)
            else:
                logger.warning(
                    f"Unknown generation type {generation_type}",
                    extra={"zone_key": zone_key},
                )

        production_data.append(
            zoneKey=zone_key,
            datetime=datetime.fromisoformat(
                quarter_hourly_source_data["DateTime"]
            ).replace(tzinfo=TIMEZONE),
            production=production,
            source=SOURCE_NAME,
        )

    return production_data.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
