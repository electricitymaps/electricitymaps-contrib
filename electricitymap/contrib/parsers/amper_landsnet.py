from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.exceptions import ParserException

SOURCE = "amper.landsnet.is"
SOURCE_URL = "https://amper.landsnet.is/generation/api/Values"


def fetch_production(
    zone_key: ZoneKey = ZoneKey("IS"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known production mix (in MW) of a given country."""
    r = session or Session()
    if target_datetime is not None:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    res = r.get(SOURCE_URL)
    if not res.ok:
        raise ParserException(
            parser="amper_landsnet.py",
            message=f"Failed to fetch {SOURCE}: {res.status_code}, err: {res.text}",
            zone_key=zone_key,
        )

    obj = res.json()
    mix = ProductionMix(
        geothermal=obj["geothermal"],
        hydro=obj["hydro"],
        oil=obj["oil"],
    )
    breakdowns = ProductionBreakdownList(logger=logger)
    breakdowns.append(
        zoneKey=zone_key,
        datetime=datetime.fromisoformat(obj["timestamp"]).replace(
            tzinfo=ZoneInfo("Atlantic/Reykjavik")
        ),
        production=mix,
        source=SOURCE,
    )
    return breakdowns.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
