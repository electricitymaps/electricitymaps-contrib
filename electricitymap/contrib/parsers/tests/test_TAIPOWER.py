from requests_mock import GET

from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.TAIPOWER import PRODUCTION_URL, fetch_production


def test_production(adapter, session, snapshot):
    with open(
        "electricitymap/contrib/parsers/tests/mocks/TAIPOWER/genary.json", "rb"
    ) as mock_file:
        adapter.register_uri(
            GET,
            PRODUCTION_URL,
            content=mock_file.read(),
        )

    assert snapshot == fetch_production(
        zone_key=ZoneKey("TW"),
        session=session,
    )
