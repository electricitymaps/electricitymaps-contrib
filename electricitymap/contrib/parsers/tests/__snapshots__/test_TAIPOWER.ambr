# serializer version: 1
# name: test_production
  list([
    dict({
      'capacity': dict({
        'biomass': 60.7,
        'coal': 13108.6,
        'gas': 17871.300000000003,
        'geothermal': 7.2,
        'hydro': 2123.8999999999996,
        'nuclear': 951.0,
        'oil': 1591.9,
        'solar': 14053.699999999999,
        'unknown': 631.4,
        'wind': 2873.8,
      }),
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 5, 14, 15, 30, tzinfo=zoneinfo.ZoneInfo(key='Asia/Taipei')),
      'production': dict({
        'biomass': 31.9,
        'coal': 8637.7,
        'gas': 15355.7,
        'geothermal': 2.3,
        'hydro': 1157.9,
        'nuclear': 841.4,
        'oil': 419.8,
        'solar': 6775.989,
        'unknown': 1791.3,
        'wind': 181.9,
      }),
      'source': 'taipower.com.tw',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': 5.2,
        'hydro': 474.0,
      }),
      'zoneKey': 'TW',
    }),
  ])
# ---
